worlds/checksmate/Rules.py

```diff
@@ -0,0 +1,248 @@
+    difficulty = 1.0
+    if opts.fairy_chess_army.value == opts.fairy_chess_army.option_stable:
+        difficulty *= 1.05
+    if opts.fairy_chess_pawns.value != opts.fairy_chess_pawns.option_vanilla:
+        difficulty *= 1.05
+    if opts.fairy_chess_pawns.value == opts.fairy_chess_pawns.option_mixed:
+        difficulty *= 1.05
+    fairy_pieces = opts.fairy_chess_pieces_configure.value
+    if opts.fairy_chess_pieces.value == opts.fairy_chess_pieces.option_fide:
+        fairy_pieces = [0]
+    elif opts.fairy_chess_pieces.value == opts.fairy_chess_pieces.option_betza:
+        fairy_pieces = [0, 1, 2, 3]
+    elif opts.fairy_chess_pieces.value == opts.fairy_chess_pieces.option_full:
+        fairy_pieces = [0, 1, 2, 3, 4, 5]
+    difficulty *= 0.99 + (0.01 * len(fairy_pieces))
+    # difficulty *= 1 + (0.025 * (5 - self.options.max_engine_penalties))
+
+    if opts.difficulty.value == opts.difficulty.option_daily:
+        difficulty *= 1.1  # results in, for example, the 4000 checkmate requirement becoming 4400
+    if opts.difficulty.value == opts.difficulty.option_bullet:
+        difficulty *= 1.2  # results in, for example, the 4000 checkmate requirement becoming 4800
+    if opts.difficulty.value == opts.difficulty.option_relaxed:
+        difficulty *= 1.35  # results in, for example, the 4000 checkmate requirement becoming 5400
+    return difficulty
+
+
+def determine_material(opts: CMOptions, base_material: int):
+    difficulty = determine_difficulty(opts)
+    material = base_material * 100 * difficulty
+    material += progression_items["Play as White"].material * difficulty
+    return material + determine_relaxation(opts)
+
+
+def determine_min_material(opts: CMOptions):
+    return determine_material(opts, 41)
+
+
+def determine_max_material(opts: CMOptions):
+    return determine_material(opts, 46)
+
+
+def determine_relaxation(opts: CMOptions) -> int:
+    if opts.difficulty.value == opts.difficulty.option_bullet:
+        return 120
+    if opts.difficulty.value == opts.difficulty.option_relaxed:
+        return 240
+    return 0
+
+
+def meets_material_expectations(state: CollectionState,
+                                material: int, player: int, difficulty: float, absolute_relaxation: int) -> bool:
+    # TODO: handle other goals
+    target = (material * difficulty) + (absolute_relaxation if material > 90 else 0)
+    return state.prog_items[player]["Material"] >= target
+
+
+def meets_chessmen_expectations(state: CollectionState,
+                                count: int, player: int, pocket_limit_by_pocket: int) -> bool:
+    chessmen_count = state.count_group("Chessmen", player)
+    if pocket_limit_by_pocket == 0:
+        return chessmen_count >= count
+    pocket_count = ceil(state.count("Progressive Pocket", player) / pocket_limit_by_pocket)
+    return chessmen_count + pocket_count >= count
+
+
+def set_rules(multiworld: MultiWorld, player: int, opts: CMOptions):
```

Comment:
You should be able to restructure this to just take the world object and then grab player and options from that. Which also means things like `multiworld.get_location(name, player)` could just be `world.get_location(name)`

Reply:
+1 for this design pattern, passing `world: ChecksMateWorld` and then using `world.multiworld` and `world.player` is much preferred nowadays

Reply:
Hey, it looks like I also reference `completion_condition`, in the style recommended by the `world_api` documentation. Do you know if an individual world can define its `completion_condition`?

https://github.com/ArchipelagoMW/Archipelago/blob/85a713771b98aca84c900f809b0ccf41d1002d6d/docs/world%20api.md?plain=1#L645

Reply:
They don't/can't right now. So you'd just have `self.multiworld.completion_condition[self.player]`

Reply:
(Or `world.multiworld.completion_condition[world.player]`, depending on whether it's a member function of the world class or a function like this that gets passed the world)